[WebProfilerBundle] Render file links for twig templates (Fix #24355)

pquerner opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Fixed tickets | #24355 and #24218
| License       | MIT

PR #24236 introduced a bug for Windows machines and this PR attempts to fix it.
See related discussions in tickets #24355 and #24218 (and starting from comment https://github.com/symfony/symfony/pull/24236#issuecomment-332780218)


//Edit
I'm not sure why so many commits are attached here, this is the commit I wanted to PR:

https://github.com/pquerner/symfony/commit/2aeed1f8bff716607db796d625bd2d25af3c722f

Commit:
Fix URL generation by escaping file path for JS

This fixes #24355

stof commented:
This is an invalid fix:

- it only fixes a single place in the codebase for your own usecase of using a javascript expression as link. All other places are still broken (links to PHP files for controllers for instance).
- it has the potential of breaking things for people using something else than JS code (where JS escaping might hurt).

The issue you have is that your own pattern expects extra escaping. This should be handled inside the `file_link` implementation, at the time doing the replacement. this would mean doing a custom implementation of the FileLinkFormatter and replacing the service.

pquerner commented:
I'm not sure why I'm getting javascript code and [others dont.](https://github.com/symfony/symfony/pull/24236#issuecomment-332787313)
So I thought I could fix it right there in the template. 
Param "url" for the escape twig function does not work.

But I guess you are correct, best would be to fix it somewhere here
```\Symfony\Bridge\Twig\Extension\CodeExtension::getFileLink```

Commit:
Revert "Fix URL generation by escaping file path for JS"

This reverts commit 2aeed1f8bff716607db796d625bd2d25af3c722f.

Commit:
Reattempt to fix #24356
by checking if we run on Windows and $file contains a backslash char
if we do, we apply PHPs addslashes function to $file

stof commented:
@pquerner you have this either in your FrameworkBundle config (`framework.ide`) or in your php.ini for `debug.file_link_format`.

stof commented on src/Symfony/Bridge/Twig/Extension/CodeExtension.php:
```diff
@@ -202,6 +202,10 @@ public function formatFile($file, $line, $text = null)
     {
+        //If on Windows, we must escape $file if it contains backslash char
+        if('\\' === DIRECTORY_SEPARATOR && FALSE !== strpos($file, '\\')) {
+            $file = addslashes($file);
```
As said before, this is a nogo as most cases don't want such escaping. You will probably break it for all other users.

pquerner commented:
True, ```xdebug.file_link_format``` gives me that URL. ._.
```javascript:var rq = new XMLHttpRequest(); rq.open('GET', 'http://localhost:8091?message=%f:%l', true); rq.send(null);```

stof commented:
btw, XDebug would suffer from the same issue, as it would not escape the path for usage inside a JS string either.

But why do you need to use XHR to access the URL ? Why not using `http://localhost:8091?message=%f:%l` as link ?

pquerner commented:
I dont _need_ to use XHR, but I cannot get it without javascript. I disabled Xdebug, disabled the Xdebug config in PHP.ini (file_link_format), cleared cache and it still outputs the javascript stuff. I have no idea what I have done to my env. :/

stof commented:
@pquerner have you restarted PHP-FPM (or Apache if you use mod_php) ? Otherwise, the php.ini change won't be taken into account (PHP-FPM loads the config when starting)

stof commented:
and have you cleared the prod cache ? The CLI runs in the dev environment by default, not the prod one. But the code snupped you gave is from the prod environment

pquerner commented:
The code is there in prod and dev environment. Yes, I have added ```env``` parameter to ```cache:clear```. I tried running in prod env once, but then I memorized the profiler is unavailable then.

pquerner commented:
Yes, I have restarted my apache server (I develop things locally via XAMPP, it has a control panel where you can start/stop the server easily) multiple times. Tried different browsers too, no luck.

ro0NL commented:
Are you using this maybe? https://github.com/computerminds/parrot/blob/edfa1b99e629a090129908d34eee40ad7930765b/parrot-config/php/parrot-base.ini#L62

pquerner commented:
Yes, I already found out it was the XDebug config. https://github.com/symfony/symfony/pull/24356#issuecomment-332799465

Ive been reconfiguring my local dev maschine to use a different XAMPP version and now it works like expected. Sorry for the big trouble, guys. 
Nothing needs to be changed for your original PR. 

_I feel stupid_ :(
